Application: Project Aurras MVP - Phase 2

MuhammedIrfan opened the pull request:
### Project Abstract

Aurras is a middleware for the web3 ecosystem that acts as an event processor and a low code workflow orchestration platform.

This is an application for a follow-up grant: https://github.com/w3f/Grants-Program/pull/163

### For which grant [level](https://github.com/w3f/Grants-Program#level_slider-levels) are you applying? 
- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $50,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for > $100k Web3 Foundation Council approval)

### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [x] A BTC or Ethereum (DAI/USDT) address for the payment of the milestones is provided inside the application.
- [x] I have read and acknowledged the [terms and conditions](https://github.com/w3f/Grants-Program/blob/master/docs/T&Cs.md).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process).

alxs commented:
Thanks for the application to continue working on this @MuhammedIrfan. I have a few questions.

- Are you talking to anyone who might be interested in using Aurras? Seeing as we've already provided a fair amount of funding for this project, it would be great to have a third party vouch for its usefulness.
- What are your plans to fund its continued development after this grant?
- Could you expand on the deliverables for M2 'methods for the flow design' and 'methods for the task struct based on the type of it'.
- Out of curiosity, how did you find out about Tackle Box?
- Since I can't seem to find it described anywhere: what exactly will the tasks that workflows are composed of be?
- Will the project be a functional MVP by the end of this grant? If so, would it maybe make sense to add some user stories as deliverables or could you provide a few examples of what will be possible after completion?

MuhammedIrfan commented:
> Thanks for the application to continue working on this @MuhammedIrfan. I have a few questions.

> * Are you talking to anyone who might be interested in using Aurras? Seeing as we've already provided a fair amount of funding for this project, it would be great to have a third party vouch for its usefulness.

We have conducted a small product research survey before starting the MVP and have identified potential need for such a product in the community. After the MVP is completely functional we intend to reach out to the community for further validation, improvement and suggestions for the product. As a future plan we have plans to include user interface to manage the workflow.

> * What are your plans to fund its continued development after this grant?

Our vision for this project is to be truly decentralized and an open network right from the very beginning. Hence in order to make our vision come true, we are required to fund the project through personal and various grant opportunities that we are exploring and other funding alternatives that does not compromise our vision.

> * Could you expand on the deliverables for M2 'methods for the flow design' and 'methods for the task struct based on the type of it'.

We have updated the grant application with more details

> * Out of curiosity, how did you find out about Tackle Box?

While researching about code generator and yaml parser we came across tackle box and upon reaching the team they provided a great support for us. We have identified this tool is compatible for aurras as it delivers a flexible way to both generate the required code and expose customization in how the workflows are built within the same tool. 

We originally connected through the w3f ecosystem over common interests in building event driven architectures for the community.

> * Since I can't seem to find it described anywhere: what exactly will the tasks that workflows are composed of be?

The tasks can be any kind of outbound call be it an http request or a smart contract call or transformation needed to reshape the data before it is output the workflow terminator.

> * Will the project be a functional MVP by the end of this grant? If so, would it maybe make sense to add some user stories as deliverables or could you provide a few examples of what will be possible after completion?

Yes, We have targeted in such a way that both phases of MVP will be functional. MVP for phase 1 is functional where we can get push notification for balance changes. In phase 2 we intend to extend it for generic use cases. We’ll include some user stories as deliverables

alxs commented:
Thanks for the extensive answers. Sounds good - consider applying for treasury funding after this grant, as you will already have an MVP and this will help you promote the project. Also, a few follow-up questions:

- You quote 4 FTEs but you seem to be the sole developer of the project so far. Are you aware that FTEs represent the number of people working on the project at the same time?
- Will users need to create tasks manually, will default ones be provided, or how do you imagine this?
- It would be great if you could describe somewhere in the application exactly what kind of tasks will exist/it will be possible to create after this grant. User stories are great, although I think it would make a lot more sense to include those now with the application, and instead deliver good user documentation with the last milestone.

MuhammedIrfan commented:
> Thanks for the extensive answers. Sounds good - consider applying for treasury funding after this grant, as you will already have an MVP and this will help you promote the project. Also, a few follow-up questions:

Sure @alxs, Thank you for your guidance.

> * You quote 4 FTEs but you seem to be the sole developer of the project so far. Are you aware that FTEs represent the number of people working on the project at the same time?

Yes, I am aware of it. For the MVP Phase 1 we followed a centralized approach wherein I have reviewed the code and committed, But this time we have planned to push the code with original contributors.

> * Will users need to create tasks manually, will default ones be provided, or how do you imagine this?

We will be updating the existing tasks that cover the use case of notifying the user for balance updates (Balance Filter, Push Notification) to follow the generic specification, Tasks can be any serverless functions that can be created using NodeJS, Go, Java, Scala, PHP, Python, Ruby, Ballerina, .NET, and Rust. Users can create the tasks following the specification using one of the supported languages. Tasks can be considered as any action that needs to be performed as part of the workflow. Considering the below request we are planning to include tasks to cover the user story that we are planning to implement in this grant.

> * It would be great if you could describe somewhere in the application exactly what kind of tasks will exist/it will be possible to create after this grant. User stories are great, although I think it would make a lot more sense to include those now with the application, and instead deliver good user documentation with the last milestone.

We are planning to implement a user story and update the grant application for the same, We have been maintaining the user documentation in https://docs.aurras.hugobyte.com/ and we will be updating the documentation alongside the new implementations

alxs commented:
Sounds good, let me know when you're done updating the application.

MuhammedIrfan commented:
> Sounds good, let me know when you're done updating the application.

Sure @alxs

MuhammedIrfan commented:
@alxs We have planned to implement polkadot validator payout as user story, We are preparing the deliverables for implementing the use case and the proposal will be updated soon with the details.

MuhammedIrfan commented:
@alxs We have updated the application kindly review it.

alxs reviewed (changes requested):
Thanks for the updates. Please note that these don't represent user stories - not sure if that was intentional. Personally, I think they would be highly useful in depicting what functionality the project would provide after this grant. As I understand it now, out of the box it would allow one to automate validator payouts and to push this to Kafka? Besides, could you please integrate the description of tasks from your last comment into the application, ideally into a deliverable?

MuhammedIrfan commented:
@alxs To better understand your suggestions, I would like to propose a meeting. Please let me know of a time slot of your convenience.

alxs commented:
That'd be great. I've sent you an email.

MuhammedIrfan commented:
@alxs The ambiguity created by the interchanged usage of 'action' and 'task' has been resolved in the updated proposal. As suggested, user stories have been added to the deliverables. Kindly review. We appreciate the support you and W3F have been providing.

alxs commented:
Thanks a lot for the added info. The application looks good and I'd say in its current state it's ready to be moved to the review stage. However, I feel like the distinction between OpenWhisk's capabilities and what you're building was crucial and in that regard, the application currently seems really expensive to me. Would you be willing to reduce the price?

By the way, you currently quote 4 FTEs but you only list 3 team members in management positions, who likely won't be working on this full time. Could you add some information on who will actually be working on the project?

MuhammedIrfan commented:
@alxs Thank you for reviewing the application.

> Thanks a lot for the added info. The application looks good and I'd say in its current state it's ready to be moved to the review stage. However, I feel like the distinction between OpenWhisk's capabilities and what you're building was crucial and in that regard, the application currently seems really expensive to me. Would you be willing to reduce the price?

Currently, we have estimated the cost at $20/hr ie 268 Person Days * 8 Hrs * $20 = 42,880. As the implementation of Polkadot validator payout wasn't in our initial planning we can work out by removing the Polkadot validator payout deliverables, Would you be open to that?

> By the way, you currently quote 4 FTEs but you only list 3 team members in management positions, who likely won't be working on this full time. Could you add some information on who will actually be working on the project?

Sure @alxs, We will update the application with the team details and their respective role in the project

Noc2 reviewed (changes requested):
Thanks for the application and I’m generally happy to support it. But I have two questions: 

- You still say that you have three team members, but you have 4 FTE per milestone. Is someone else working on your project? Otherwise could you potentially reduce the price and change it to 3 FTE? 
- Do you think it makes sense to integrate a tutorial or something similar, that makes it easier for others to adapt the project for their own events?

MuhammedIrfan commented:
@Noc2 Thank you for reviewing the application.

> Thanks for the application and I’m generally happy to support it. But I have two questions:
> 
> * You still say that you have three team members, but you have 4 FTE per milestone. Is someone else working on your project? Otherwise could you potentially reduce the price and change it to 3 FTE?

We have updated the profile with team details.

> * Do you think it makes sense to integrate a tutorial or something similar, that makes it easier for others to adapt the project for their own events?

Definitely, We have updated the deliverables to include tutorials

Noc2 reviewed (approved):
Thanks for the update. I’m happy to go ahead with it.

MuhammedIrfan commented:
> Thanks for the update. I’m happy to go ahead with it.

Thank you @Noc2

takahser commented on applications/project_aurras_mvp_phase_2.md:
```diff
@@ -0,0 +1,239 @@
+### Project Details
+
+#### Architecture
+![](https://storage.googleapis.com/hugobyte-2.appspot.com/architecture.jpeg)
```
The top left graphic represents a substrate-based blockchain, right?
![image](https://user-images.githubusercontent.com/5393704/161025545-b62e5973-0951-4209-b15b-6d6555aae752.png)

It might be useful to add a corresponding label here, so it's clear for everybody what it represents.

MuhammedIrfan replied:
We have added label to the image

takahser commented on applications/project_aurras_mvp_phase_2.md:
```diff
@@ -0,0 +1,239 @@
+
+1. [Apache OpenWhisk](https://openwhisk.apache.org) is an open source, distributed Serverless platform that executes functions (fx) in response to events at any scale. OpenWhisk manages the infrastructure, servers and scaling using Docker containers. The OpenWhisk platform supports a programming model in which developers write functional logic (called Actions), in any supported programming langua ...
+2. [Apache Kafka](https://kafka.apache.org) is an open-source distributed event streaming platform. Kafka works well as a replacement for a more traditional message broker. Message brokers are used for a variety of reasons (to decouple processing from data producers, to buffer unprocessed messages, etc.). In comparison to most messaging systems, Kafka has better throughput, built-in partitioning, ...
+3. [Tackle box](https://robcxyz.github.io/tackle-box/) is a declarative DSL for building modular utilities and code generators. Tool is plugins based and can easily be extended by writing additional hooks or importing external providers creating a web of interoperable CLIs.
```
By "Tool is plugins based" do you mean "the tool (tackle box) is plugin-based"?

MuhammedIrfan replied:
Yes the tool is tackle box we will update the application to remove this confusion, Thank you for pointing it out.

takahser commented on applications/project_aurras_mvp_phase_2.md:
```diff
@@ -0,0 +1,239 @@
+- **Full-time equivalent (FTE):** 4
+- **Total Costs:** 42880 USD 
+
+### Milestone 1 — Workflow Composer - Part 1
```
Are you going to deliver docker files that allow for running your deliverables? If so, it would be nice if you could add them to the milestone deliveries. See also our [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md#development-roadmap-nut_and_bolt):

> Deliverables 0a-0d are mandatory for all milestones

MuhammedIrfan replied:
Sure takahser we will update the deliverables.

takahser replied:
@MuhammedIrfan are you delivering docker files only for this delivery or also M1 and M3?

MuhammedIrfan replied:
@takahser we will have it only for M2 as its the only executable left without docker file, Rest of the docker files has been delivered as part of phase 1.

takahser commented on applications/project_aurras_mvp_phase_2.md:
```diff
@@ -0,0 +1,239 @@
+
+#### Components
+
+**1.Event Sources (Substrate Event Feed)**  
```
```suggestion
**1. Event Sources (Substrate Event Feed)**  
```

takahser commented on applications/project_aurras_mvp_phase_2.md:
```diff
@@ -0,0 +1,239 @@
+
+The indexer will use a NoSQL database to store the last indexed block and metadata. Using the substrate indexer as our event source will help us in solving many challenges ranging from fault tolerance to scalability to uniquely identifying multiple events with the same payload within a single block. The indexer will be connected to a substrate-based chain and RocksDB where the chain data is store ...
+
+**2.Event Trigger Manager**  
```
```suggestion
**2. Event Trigger Manager**  
```

takahser commented on applications/project_aurras_mvp_phase_2.md:
```diff
@@ -0,0 +1,239 @@
+
+Under the hood, a feed action is invoked with create lifecycle, which accepts the mandatory parameters the lifecycle, auth, trigger name, and other optional parameters of the event source. The feed action invokes the related actions of creating the entry in the database, adding to the Kafka consumer group, etc. The next component in the event trigger manager is a persistent connection to Kafka wh ...
+
+**3.Workflow Composer**  
```
```suggestion
**3. Workflow Composer**  
```

takahser commented on applications/project_aurras_mvp_phase_2.md:
```diff
@@ -0,0 +1,239 @@
+
+Workflow configuration comprises the endpoint URL for workflow tasks, the sequence of processing tasks, and argument structure. Arguments must match the task input parameters.
+ 
+**4.Web API Gateway and Backend Service**   
```
```suggestion
**4. Web API Gateway and Backend Service**   
```

takahser commented on applications/project_aurras_mvp_phase_2.md:
```diff
@@ -0,0 +1,239 @@
+The indexer will use a NoSQL database to store the last indexed block and metadata. Using the substrate indexer as our event source will help us in solving many challenges ranging from fault tolerance to scalability to uniquely identifying multiple events with the same payload within a single block. The indexer will be connected to a substrate-based chain and RocksDB where the chain data is store ...
+
+**2.Event Trigger Manager**  
+Event trigger manager is composed of multiple OpenWhisk actions, including a database to store trigger URLs and their respective auths, the Kafka provider, consumer and producer. Once the event manager receives an event from the event feed, this data is produced to a topic. The feed action in the trigger manager lets the user hook into the system. That is,  once an event is indexed to a particula ...
```
```suggestion
Event trigger manager is composed of multiple OpenWhisk actions, including a database to store trigger URLs and their respective auths, the Kafka provider, consumer and producer. Once the event manager receives an event from the event feed, this data is produced to a topic. The feed action in the trigger manager lets the user hook into the system. That is, once an event is indexed to a particular topic, it can invoke a particular action. While creating the workflow, users can choose the event trigger as feed and provide necessary parameters from which chain it should be listening to and from which block it should start listening.
```

takahser reviewed (changes requested):
@MuhammedIrfan thanks for your application, looks interesting! I left you some comments above ⬆️

MuhammedIrfan commented on applications/project_aurras_mvp_phase_2.md:
```diff
@@ -0,0 +1,239 @@
+**1.Event Sources (Substrate Event Feed)**  
+This is the source for events from the blockchain and the events will be sent to the OpenWhisk system. The MVP event source will be implemented using [polkadot-js/api](https://github.com/polkadot-js/api). For the actual implementation, we will be using a custom version of the substrate archive. Substrate archive will be indexing the blocks and an event actor will be decoding the events and data f ...
+
+The indexer will use a NoSQL database to store the last indexed block and metadata. Using the substrate indexer as our event source will help us in solving many challenges ranging from fault tolerance to scalability to uniquely identifying multiple events with the same payload within a single block. The indexer will be connected to a substrate-based chain and RocksDB where the chain data is store ...
```
This has not been decided yet as it isn't part of the MVP, we will be coming to conclusion at a later point.

Noc2 reviewed (approved):
I’m still happy to go ahead with it.

takahser reviewed (approved):
@MuhammedIrfan thanks a lot for the updates, at this point I'm happy to support this project as well.

MuhammedIrfan commented:
> @MuhammedIrfan thanks a lot for the updates, at this point I'm happy to support this project as well.

Thank you for your support @takahser

SBalaguer commented:
Thanks @MuhammedIrfan for the proposal! Very happy to see you around. I'm overall happy to approve this proposal, the discussions before were very helpful. One thing that I would encourage though to think of is how to get the community engaged in the future to use this as part of their implementations. Looking forward to the deliveries!

MuhammedIrfan commented:
> Thanks @MuhammedIrfan for the proposal! Very happy to see you around. I'm overall happy to approve this proposal, the discussions before were very helpful. One thing that I would encourage though to think of is how to get the community engaged in the future to use this as part of their implementations. Looking forward to the deliveries!

Thank you very much for the support. Definitely we would be needing community support too. We are preparing a strategy to get engaged with community. Looking forward to produce best quality deliverables.

MuhammedIrfan commented:
> Thanks @MuhammedIrfan for the proposal! Very happy to see you around. I'm overall happy to approve this proposal, the discussions before were very helpful. One thing that I would encourage though to think of is how to get the community engaged in the future to use this as part of their implementations. Looking forward to the deliveries!

@alxs @Noc2 @takahser @SBalaguer 
Thank you very much for the support. Definitely we would be needing community support too. We are preparing a strategy to get engaged with community. Looking forward to produce best quality deliverables.

MuhammedIrfan commented:
Hi Team, This is to update you regarding the progress of our deliverables. We have stumbled upon an unexpected blocker which we are working hard to resolve soon, Hence we would require to push the 1st milestone deliverables for 2 weeks. Once again I would like to thank you for giving us the opportunity to serve the Web 3 ecosystem.

takahser commented:
@MuhammedIrfan thanks for the update, that works for us.
Looking forward to your M1 delivery. 👍

alxs commented:
Checking in again @MuhammedIrfan. If you're not planning to deliver within the next two weeks, could you please submit an [amendment](https://github.com/w3f/Grants-Program#changes-to-a-grant-after-approval) to update the timeline?

MuhammedIrfan commented:
> Checking in again @MuhammedIrfan. If you're not planning to deliver within the next two weeks, could you please submit an [amendment](https://github.com/w3f/Grants-Program#changes-to-a-grant-after-approval) to update the timeline?

@alxs Sure, I will make amendments to proposal, Thank you for your support.

alxs commented:
@MuhammedIrfan friendly reminder

keeganquigley commented:
Hi @MuhammedIrfan Are you still planning to submit an amendment to your proposal? Just a reminder that if we don't hear an update from you after two weeks, we will submit a PR to terminate the grant due to inactivity. Thanks!

MuhammedIrfan commented:
> Hi @MuhammedIrfan Are you still planning to submit an amendment to your proposal? Just a reminder that if we don't hear an update from you after two weeks, we will submit a PR to terminate the grant due to inactivity. Thanks!

I sincerely apologize for the delay. Kindly find the amendment PR https://github.com/w3f/Grants-Program/pull/1402

MuhammedIrfan commented:
Hi @keeganquigley, We are in the process of final delivery, We are currently doing the testing we will be able to submit the milestone delivery early next week. Thank you.
